appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockChartData.java

```diff
@@ -0,0 +1,405 @@
+        super.onSelectedChange(selected);
+        removeStyleDependentName("selected"); // Force remove styling
+    }
+
+    @Override
+    public void onRemoved() {
+        super.onRemoved();
+        setSourceProperty(""); // Unset the Source property to remove listener references
+        chartDataModel.removeDataSeriesFromChart(); // Remove the Data Series from the Chart
+        refreshChart();
+    }
+
+    /**
+     * Sets the current elements of the Chart Data component.
+     *
+     * @param text  new elements text
+     */
+    private void setElementsFromPairsProperty(String text){
+        currentElements = text;
+
+        chartDataModel.setElements(currentElements);
+    }
+
+    /**
+     * Sets the Source property of the Chart Data component.
+     * Responsible for showing/hiding properties according to the
+     * attached Source component type.
+     *
+     * @param source  Name of the new Source component attached
+     */
+    private void setSourceProperty(String source) {
+        // If a Data Source was previously assigned and is of type DataFile,
+        // de-attach this Data component from it
+        if (dataSource instanceof MockDataFile) {
+            ((MockDataFile)dataSource).removeDataFileChangeListener(this);
+        }
+
+        // Get the newly attached Source component
+        dataSource = editor.getComponents().getOrDefault(source, null);
+
+        changeSourcePropertiesVisibility();
+
+        // If the Data Source is now null, set back the
+        // currentElements property.
+        if (dataSource == null) {
+            onPropertyChange(PROPERTY_PAIRS, currentElements);
+        }
+
+        // If the component is currently selected, re-select it to refresh
+        // the Properties panel.
+        if (isSelected()) {
+            onSelectedChange(true);
+        }
+    }
+
+    /**
+     * Sets the Data File X Column property of the Chart Data component.
+     * After setting the property, the data is then re-imported
+     * (if possible)
+     * @param column  new column name
+     */
+    private void setDataFileXColumnProperty(String column) {
+        // The X column is the first element of the dataFileColumns list
+        dataFileColumns.set(0, column);
+        updateDataFileData();
+    }
+
+    /**
+     * Sets the Point Shape property to the Chart Data component,
+     * provided that the Data Model is a ScatterChartDataModel.
+     * If that is indeed the case, the point shape of the Scatter
+     * Chart Data Model is then changed.
+     * @param newValue  New value of the Point Shape (String)
+     */
+    private void setPointShapeProperty(String newValue) {
+        int pointShape = Integer.parseInt(newValue);
+
+        // Only change the point shape of the Model if it is of
+        // type ScatterChartDataModel (since only that model supports
+        // the changing of the Point Shape)
+        if (chartDataModel instanceof MockScatterChartDataModel) {
```

Comment:
@ewpatton I will refactor it to the suggested approach, however, we cannot yet support this for Line Charts, since the MPAndroidChart library only supports circles for the Line Chart (and different shapes are only supported for the Scatter Chart).